---
title: "Preparing data for shiny app"
output: html_notebook
---

# Setup

```{r setup, include = FALSE}
knitr::opts_chunk$set(cache = TRUE, dpi = 300)
```

```{r message = FALSE, warning = FALSE}
# for unified and processed RNA-seq data
library(recount3)
# to normalize the RNA-seq data 
library(DESeq2) 
# to look at the data
library(tidyverse)
# to visualize the mutation data
library(maftools)
```

# Getting normalized count data

```{r message=FALSE, warning=FALSE}
rse_gene <- 
  create_rse(
    subset(
      available_projects(),
      project == "LGG" & project_type == "data_sources"
    )
  )
```

```{r}
assay(rse_gene, "counts") <- 
  transform_counts(rse_gene)
```

```{r}
normalized_counts <- 
  vst(assays(rse_gene)$counts)
```

```{r}
sample_sheet <-
  colData(rse_gene) %>%
  data.frame() %>%
  rownames_to_column("sample_id")
```


```{r}
sample_sheet_one <-
  sample_sheet %>%
  filter(tcga.cgc_sample_sample_type == "Primary Tumor") %>%
  mutate(patient_id = str_extract(tcga.tcga_barcode, 
                                "[^-]{4}-[^-]{2}-[^-]{4}")) %>%
  group_by(patient_id) %>%
  # this is quick, but dirty way to take just one repeat per patient
  sample_n(1) 

# select only the samples we want to keep
normalized_counts <-
  normalized_counts[ ,  sample_sheet_one$sample_id]

# change rownames to nice patient ids
colnames(normalized_counts) <- sample_sheet_one$patient_id
```

```{r}
ensembl_ids <- 
  rownames(normalized_counts) %>%
  str_remove("\\.[0-9]*")

mart <- 
  biomaRt::useMart("ensembl", dataset = "hsapiens_gene_ensembl")

ensembl_to_hgnc <-
  biomaRt::getBM(attributes = c('ensembl_gene_id', 'hgnc_symbol'), 
                 mart = mart) %>%
  filter(ensembl_gene_id %in% ensembl_ids) %>%
  group_by(ensembl_gene_id) %>%
  summarise(hgnc_symbol = paste(unique(hgnc_symbol)[unique(hgnc_symbol) != ""],
                                collapse = ",")) %>%
  mutate(hgnc_symbol = ifelse(hgnc_symbol == "", 
                              ensembl_gene_id,
                              hgnc_symbol)) %>%
  column_to_rownames("ensembl_gene_id") 

rownames(normalized_counts) <-
  ensembl_to_hgnc[ensembl_ids,]
```
```{r}
normalized_counts[1:5, 1:5]
```

# mutation data

```{r maf, message=FALSE, warning=FALSE, error=FALSE, results='hide'}
maf <-
  GDCquery_Maf("LGG", pipelines = "mutect2") %>%
  read.maf(verbose = TRUE)

# now that GDC works again, we don't need this
# 
# tryCatch(maf <- tcgaLoad(study = "LGG"), 
#          error = function(e) {
#            print(paste(rep("#", 50), collapse = ""))
#            print(paste0("# ERROR! Read the message below!", 
#                         paste(rep(" ", 17), collapse = ""),
#                         "#"))
#            print(paste(rep("#", 50), collapse = ""))
#            print(e)
#            print(paste("If you're seeing this message you probably don't have",
#                        "maftools package loaded, or have an older version.", 
#                        "This function is available with v2.8.",
#                        "Install the new version of maftools package with",
#                        "`BiocManager::install('PoisonAlien/maftools')`", 
#                        "and try again!"))
#            })
```


# Joining the data 

Check if we have the necessary data to create the plots.

```{r}
# which genes are we intersted in?
genes_of_interest <- c("IDH1", "TP53")

# this will allow us to distinguish between no information about mutation
# and no mutation
patient_muts <-
  maf@data %>%
  mutate(patient_id = str_extract(Tumor_Sample_Barcode, 
                                  "[^-]{4}-[^-]{2}-[^-]{4}")) %>%
  pull(patient_id) %>%
  unique()

expression_vals <-
  normalized_counts[genes_of_interest,] %>%
  as.data.frame() %>%
  rownames_to_column("Hugo_Symbol") %>%
  pivot_longer(names_to = "patient_id", 
              values_to = "norm_expression",
              cols = -Hugo_Symbol)

mutation_data <-
  maf@data %>%
  filter(Hugo_Symbol %in% genes_of_interest) %>%
  mutate(patient_id = str_extract(Tumor_Sample_Barcode, 
                                  "[^-]{4}-[^-]{2}-[^-]{4}")) %>%
  dplyr::select(patient_id, Hugo_Symbol, VARIANT_CLASS)

expr_mut_df <-
  full_join(expression_vals, mutation_data) %>%
  # introduce WT for samples with no mutation in the gene
  mutate(VARIANT_CLASS = as.character(VARIANT_CLASS),
         VARIANT_CLASS = ifelse(is.na(VARIANT_CLASS),
                                         yes = ifelse(patient_id %in% patient_muts,
                                                yes = "WT",
                                                no = "NA"), 
                                         no = VARIANT_CLASS),
         VARIANT_CLASS = factor(VARIANT_CLASS, 
                                levels = c("WT", 
                                           "SNV",
                                           "deletion", "insertion",
                                           "NA")))

expr_mut_df %>%
  ggplot(aes(VARIANT_CLASS, norm_expression, color = VARIANT_CLASS,
             shape = VARIANT_CLASS)) +
  geom_jitter() +
  geom_boxplot(color = "black", width = 0.3, 
               alpha = 0.7, outlier.shape = NA) +
  facet_wrap(~Hugo_Symbol) +
  theme_classic() +
  theme(legend.position = "none") +
  labs(y = "vst normalized expression",
       x = "Variant classification")
```

## Save the data

```{r}
app_data_dir <- "../interactive_app/data"
if (!dir.exists(app_data_dir)) {
  dir.create(app_data_dir)
}
saveRDS(maf@data, file.path(app_data_dir, "maf_data_df.RDS"))
saveRDS(normalized_counts, file.path(app_data_dir, "normalized_counts.RDS"))
```

